core/src/main/java/org/apache/calcite/rel/hint/NodeTypeHintStrategy.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.calcite.rel.hint;
+
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.core.TableScan;
+
+/**
+ * A hint strategy that specifies which kind of relational
```

Comment:
Actually I am not very sure how useful to apply hints on a relational node without considering its inputs. Based on my experiences, almost all relational hints would be specific to a particular node (such as a join node with specific inputs).

Reply:
Some may not attach to a specific node, like parallelism, memory

Reply:
For  parallelism and memory, we have a node type `NodeType.QUERY` which indicates that the hint would not propagate.

Reply:
Oops, maybe with the name `SET_VAR` you suggested.